### PR TITLE
tools: add policy generation tool

### PR DESCRIPTION
**Pull request by danmihai1**
Add application that infers K8s user's intentions based on user's K8s YAML file, and generates a Rego/OPA based policy for that YAML.

Fixes: #7673

### CODE REVIEW

File: src/tools/genpolicy/README.md
Comment (stevenhorsman) [inline src/tools/genpolicy/README.md:9]: I'd be interested in understanding/communicating some of the rules that genpolicy used based on input patterns, or at least some examples as the policy file that is output is pretty large (the basic pod yaml example you give it decodes to >1600 lines long) so is tricky to understand e.g. things like the container image name will be extracted and added to the list of images that are allowed in the CreateContainerRequest, or something like that?

Maybe explaining the genpolicy-settings.json file, which looks like it does some templating? - what it is and what is does would help with this?

**Review by stevenhorsman — COMMENTED** [review]
Hey Dan, thanks so much for the submission. There is a lot of code here and I'm not an expert, so I've focussed on the documentation and through to get the tool build and running myself primarily, and left some comments about how we might be able to make this a bit more accessible as there were a few hurdles I bumped into and the tool has a lot of power, so letting people understand the processing and output more feels important. Thanks!

**Comment by danmihai1** [comment]
> Hey Dan, thanks so much for the submission. There is a lot of code here and I'm not an expert, so I've focussed on the documentation and through to get the tool build and running myself primarily, and left some comments about how we might be able to make this a bit more accessible as there were a few hurdles I bumped into and the tool has a lot of power, so letting people understand the processing and output more feels important. Thanks!

Thanks for all the feedback @stevenhorsman. I have just pushed what are significant doc improvements in https://github.com/kata-containers/kata-containers/pull/8248/commits/2257a1691794e4d2ad0c7e035ce8a4435f94d7c4

File: src/tools/genpolicy/README.md
Comment (danmihai1) [reply src/tools/genpolicy/README.md]: Added that note.

File: src/tools/genpolicy/README.md
Comment (danmihai1) [reply src/tools/genpolicy/README.md:1]: Added the build steps into the new version of README.md

File: src/tools/genpolicy/README.md
Comment (danmihai1) [reply src/tools/genpolicy/README.md]: I've just added many policy details into the new genpolicy-auto-generated-policy-details.md, and advanced command line usage in genpolicy-advanced-command-line-parameters.md.

**Review by stevenhorsman — APPROVED** [review]
Thanks very much for the extra documentation Dan - it really helps. Again the code is too much for me to do a sensible review, but I've kicked the tyres with testing, so I think it's fine to go in as the initial drop of some tooling.

**Review by katexochen — CHANGES_REQUESTED** [review]
I'm also if there are any requirements for the container images used. Not sure if that's written anywhere in the documentation and I've missed it. I've tested a bunch of images, some worked, some failed:

```yaml
apiVersion: v1
kind: Pod
metadata:
  name: test
spec:
  containers:
    - name: test
      image: ${image}
```

`image="ubuntu:latest"`
```
Error: failed to generate policies: failed to generate policy for deployment/test.yml: genpolicy failed with exit code 101: thread 'main' panicked at src/registry.rs:111:17:
Failed to pull container image manifest and config - error: UnsupportedMediaTypeError(
    "application/vnd.oci.image.index.v1+json",
)
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```
`image="nixery.dev/shell/socat"`
```
Error: failed to generate policies: failed to generate policy for deployment/test.yml: genpolicy failed with exit code 101: thread 'main' panicked at src/registry.rs:91:61:
called `Result::unwrap()` on an `Err` value: Error("missing field `Env`", line: 1, column: 969)
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```

File: src/tools/genpolicy/src/policy.rs
Comment (danmihai1) [reply src/tools/genpolicy/src/policy.rs:366]: @katexochen , thanks for the feedback! This tool doesn't solve all the Confidential K8s Policy problems (yet?). For example, if you search for TODO in its source code, you'll find a bunch of other problem that were not solved yet. I think we should start somewhere and keep improving. After working on this for several months, I think that waiting for the ideal tool before the first merge is impractical.

**Comment by danmihai1** [comment]
> I'm also if there are any requirements for the container images used. Not sure if that's written anywhere in the documentation and I've missed it. I've tested a bunch of images, some worked, some failed:
> 
> ```yaml
> apiVersion: v1
> kind: Pod
> metadata:
>   name: test
> spec:
>   containers:
>     - name: test
>       image: ${image}
> ```
> 
> `image="ubuntu:latest"`
> 
> ```
> Error: failed to generate policies: failed to generate policy for deployment/test.yml: genpolicy failed with exit code 101: thread 'main' panicked at src/registry.rs:111:17:
> Failed to pull container image manifest and config - error: UnsupportedMediaTypeError(
>     "application/vnd.oci.image.index.v1+json",
> )
> note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
> ```
> 
> `image="nixery.dev/shell/socat"`
> 
> ```
> Error: failed to generate policies: failed to generate policy for deployment/test.yml: genpolicy failed with exit code 101: thread 'main' panicked at src/registry.rs:91:61:
> called `Result::unwrap()` on an `Err` value: Error("missing field `Env`", line: 1, column: 969)
> note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
> ```

Please log github issues for YAML files that are not working and attach the problematic YAML files to those issues. For these particular two problems:

1. One of my colleagues at MSFT noticed the ubuntu:latest problem a few days ago, and we are working on a solution for it. I can share the gory details in case you are curious about the root cause. Meanwhile, creating your own container image on top of an Ubuntu base container image (but using a different manifest format) might be a good enough workaround.
2. Please attach your test.yaml in github - the Env problem you mentioned sounds more related to the YAML contents, not to the container image.

File: src/tools/genpolicy/genpolicy-advanced-command-line-parameters.md
Comment (danmihai1) [reply src/tools/genpolicy/genpolicy-advanced-command-line-parameters.md:50]: I agree that caching needs future improvements, but I don't think the initial merge of this code should wait for these improvements to become available. Please create github issues for such ideas for improvements.

File: src/tools/genpolicy/genpolicy-advanced-command-line-parameters.md
Comment (danmihai1) [reply src/tools/genpolicy/genpolicy-advanced-command-line-parameters.md:86]: Please log an issue for the future.

File: src/tools/genpolicy/Cargo.toml
Comment (ChengyuZhu6) [inline src/tools/genpolicy/Cargo.toml:37]: > 1. One of my colleagues at MSFT noticed the ubuntu:latest problem a few days ago, and we are working on a solution for it. I can share the gory details in case you are curious about the root cause. Meanwhile, creating your own container image on top of an Ubuntu base container image (but using a different manifest format) might be a good enough workaround.

Hey @danmihai1, I noticed the problem with ubuntu:latest, which was caused by the manifest type. We should handle manifests with the mediatype “application/vnd.oci.image.manifest.v1+json” when pulling them from the registry. This issue had been solved in the [oci-distribution](https://github.com/krustlet/oci-distribution/pull/67). Could you please change the oci-distribution repo branch to [this commit](https://github.com/krustlet/oci-distribution/pull/67/commits/0036d93e8f6eb1df4e01e60c4f1c9c77ea43461c)?”

File: src/tools/genpolicy/src/policy.rs
Comment (ChengyuZhu6) [inline src/tools/genpolicy/src/policy.rs:626]: Hey @danmihai1. Am I correct in assuming that this is an interface for sharing image layers? I am thrilled to see this feature, as I am working on integrating image sharing feature into the main branch (https://github.com/kata-containers/kata-containers/pull/7837). I would like to pass the dm-verity information of the image or layer to the guest via the policy to check the info from host is correct as expected, if possible. I would appreciate your assistance with this task.

Comment (danmihai1) [reply src/tools/genpolicy/src/policy.rs:626]: Please ping me on Slack when you are ready to discuss this. 

Yes, the dm-verity information of each container layer is being verified already when using the Policy generated by the genpolicy tool. But, the current Policy is based on Wedson's snapshotter design/implementation that we discussed several months ago. Hopefully there aren't big differences between your version and Wedson's. If there are significant differences, we'll have to modify this tool to support both.

**Comment by danmihai1** [comment]
No code or doc changes in my latest push. Just squashed the doc changes into a single commit and rebased on top of the latest main code.

File: src/tools/genpolicy/Cargo.toml
Comment (danmihai1) [reply src/tools/genpolicy/Cargo.toml]: @ChengyuZhu6 @katexochen Christmas came early for me this year and brought a new release of the oci_distribution crate - a few days ago. Therefore, I have just added to this PR @Redent0r's commit that adds support for the ubuntu:latest image by using the new oci_distribution version.

**Comment by huangyingting** [comment]
@danmihai1 I want to report an issue while I am testing confidential containers from AKS, there is no place from microsoft/kata-containers to let me open an issue, so I put it here

It seems we not able to correctly handle mountPropagation: Bidirectional or mountPropagation: HostToContainer on emptyDir volume, as a consequence, the container won't be created with error "Error: failed to create containerd task: failed to create shim task: "CreateContainerRequest is blocked by policy": unknown"

To work around the issue, for HostToContainer, we need to replace rprivate to rslave in the generated policy.

          {
            "destination": "/mnt/remote",
            "source": "^$(cpath)/$(sandbox-id)/local/remotemounts$",
            "type_": "local",
            "options": [
              "rbind",
              "rslave",
              "rw"
            ]
          }
For Bidirectional, we need to replace rprivate to rshared in the generated policy.

          {
            "destination": "/mnt/remote",
            "source": "^$(cpath)/$(sandbox-id)/local/remotemounts$",
            "type_": "local",
            "options": [
              "rbind",
              "rshared",
              "rw"
            ]
          }

To reproduce the issue, you can use https://github.com/huangyingting/sev-snp/blob/main/encfs.yaml (ignore io.katacontainers.config.agent.policy, it was tweaked already based on above findings to make it work)

**Comment by danmihai1** [comment]
> @danmihai1 I want to report an issue while I am testing confidential containers from AKS, there is no place from microsoft/kata-containers to let me open an issue, so I put it here
> 
> It seems we not able to correctly handle mountPropagation: Bidirectional or mountPropagation: HostToContainer on emptyDir volume, as a consequence, the container won't be created with error "Error: failed to create containerd task: failed to create shim task: "CreateContainerRequest is blocked by policy": unknown"

Thanks! We are tracking other bugs applicable to this app under https://github.com/kata-containers/kata-containers/issues .

**Review by fidencio — APPROVED** [review]
@danmihai1, lgtm, thanks!

I think this one should get in sooner than later, as we have been tracking the improvements in a different issue.

**Review by katexochen — APPROVED** [review]

**Pull request merged**
